pkg/rbd/rbd_attach.go

```diff
@@ -260,21 +260,38 @@ func waitForrbdImage(ctx context.Context, backoff wait.Backoff, volOptions *rbdV
 	return err
 }
 
-func detachRBDDevice(ctx context.Context, devicePath string) error {
+func detachRBDDevice(ctx context.Context, devicePath, volumeID string) error {
 	nbdType := false
 	if strings.HasPrefix(devicePath, "/dev/nbd") {
 		nbdType = true
 	}
 
-	return detachRBDImageOrDeviceSpec(ctx, devicePath, false, nbdType)
+	return detachRBDImageOrDeviceSpec(ctx, devicePath, false, nbdType, volumeID)
 }
 
 // detachRBDImageOrDeviceSpec detaches an rbd imageSpec or devicePath, with additional checking
 // when imageSpec is used to decide if image is already unmapped
-func detachRBDImageOrDeviceSpec(ctx context.Context, imageOrDeviceSpec string, isImageSpec, ndbType bool) error {
-	var err error
+func detachRBDImageOrDeviceSpec(ctx context.Context, imageOrDeviceSpec string, isImageSpec, ndbType bool, volumeID string) error {
 	var output []byte
 
+	mapperFile, mapperPath := util.VolumeMapper(volumeID)
```

Comment:
(optimization) Instead of checking this for each detach request, that will introduce unwanted luks checks for non-encrypted volumes as well, pass in a flag indicating this volume is encrypted or not, and act accordingly.

In Stage phase we can stash the encrypted state of the device into the stash file and use it appropriately to pass in the flag when Unstage calls this function. (see `stashRBDImageMetadata`)

Reply:
agree that an optimization can be done, but I would prefer to merge existing PR first as it's going on for a long time already with a follow up optimization PR

Reply:
Ack! Let's track this as well in an issue, so that it gets done. I am not going to block the patch on an optimization (or nits for that matter).

Reply:
Suggested improvements tracked in #744